addon-test-support/@ember/test-helpers/dom/focus.ts

```diff
@@ -39,51 +44,68 @@ function getClosestFocusable(
-    __blur__(previousFocusedElement, focusTarget);
-  }
-
-  // makes `document.activeElement` be `element`. If the browser is focused, it also fires a focus event
-  focusTarget.focus();
+): Promise<FocusRecord | Event | void> {
+  return Promise.resolve()
+    .then(() => {
+      let focusTarget = getClosestFocusable(element);
+
+      const previousFocusedElement =
+        document.activeElement &&
+        document.activeElement !== focusTarget &&
+        isFocusable(document.activeElement)
+          ? document.activeElement
+          : null;
+
+      // fire __blur__ manually with the null relatedTarget when the target is not focusable
+      // and there was a previously focused element
+      return !focusTarget && previousFocusedElement
+        ? __blur__(previousFocusedElement, null).then(() =>
+            Promise.resolve({ focusTarget, previousFocusedElement })
+          )
+        : Promise.resolve({ focusTarget, previousFocusedElement });
+    })
+    .then(({ focusTarget, previousFocusedElement }) => {
+      if (!focusTarget) {
+        throw new Error('There was a previously focused element');
+      }
 
-  // Firefox does not trigger the `focusin` event if the window
-  // does not have focus. If the document does not have focus then
-  // fire `focusin` event as well.
-  if (browserIsNotFocused) {
-    // if the browser is not focused the previous `el.focus()` didn't fire an event, so we simulate it
-    fireEvent(focusTarget, 'focus', {
-      bubbles: false,
-    });
+      let browserIsNotFocused = !document?.hasFocus();
 
-    fireEvent(focusTarget, 'focusin');
-  }
+      // fire __blur__ manually with the correct relatedTarget when the browser is not
+      // already in focus and there was a previously focused element
+      return previousFocusedElement && browserIsNotFocused
+        ? __blur__(previousFocusedElement, focusTarget).then(() =>
+            Promise.resolve({ focusTarget })
+          )
+        : Promise.resolve({ focusTarget });
+    })
+    .then(({ focusTarget }) => {
+      // makes `document.activeElement` be `element`. If the browser is focused, it also fires a focus event
+      focusTarget.focus();
+
+      // Firefox does not trigger the `focusin` event if the window
+      // does not have focus. If the document does not have focus then
+      // fire `focusin` event as well.
+      let browserIsFocused = document?.hasFocus();
+      return browserIsFocused
+        ? Promise.resolve()
+        : // if the browser is not focused the previous `el.focus()` didn't fire an event, so we simulate it
+          Promise.resolve()
+            .then(() =>
+              fireEvent(focusTarget as HTMLElement | SVGElement, 'focus', {
+                bubbles: false,
+              })
+            )
+            .then(() =>
+              fireEvent(focusTarget as HTMLElement | SVGElement, 'focusin')
+            )
+            .then(() => settled());
+    })
+    .catch(() => {});
```

Comment:
Should we not let the exception propagate and fail the test if an exception is thrown?

Reply:
I mention in the comment above that the thrown error is a stand-in for the early `return` statement (see L62 of original code). Given we simply want to short-circuit the Promise chain, I didn't see any purpose in processing or propagating the error. 

Additionally, given this function is being executed in a nested Promise within the `focus` function (see L155 of new code), the parent Promise chain doesn't know whether the child Promise resolved or was rejected (AFAIK). 

This should result in the same behaviour we had when this function was synchronous. It doesn't look or feel great, but it gets the job done. 

That said, I'm open to suggestions!